### PR TITLE
Added regression tests for profiles in Theia

### PR DESCRIPTION
**Pull request by deepali-hub**
Signed-off-by: Deepali Chavan <deepali.chavan@broadcom.com>

## Release Notes

Milestone:

Changelog:

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [x] `npm run vscode:prepublish` has been executed
- [x] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by crawr — COMMENTED** [review]
Thanks for picking this up. Great progress so far. I just have a few comments:

- The tests should run in *headless* mode in the pipeline so please un-comment those lines. 
- For inputs to the Add Profiles dialogs I would change it to `fakehost.com:443` or something like this
- Locating element by xpath looks like it will be very difficult to maintain. Can this be optimized? Or I would suggest moving the `locate element` functions to a separate utility library so that in case there are changes to Theia side, only those functions will be changed and the test suites would not need to be updated.
- Sometimes I get the *element is not interactable* error for `Should Delete Profile from USS`  when `DeleteUSSProfile` is not successfully added to the tree. 
- I think there should be separate tests for just *Hiding Profiles* in addition to *Deleting Profiles*
- Overall, I would suggest removing the *Adding profile for deletion* part in the `it` blocks for delete and just hide the profiles from USS, and Jobs trees. Then the remaining `TestSeleniumProfile` can be deleted via the Data Sets tree.
- Minor cosmetic comment: I would recommend sticking to camel case for [variable names](https://github.com/zowe/zlc/blob/master/process/CODING_STYLE.md#naming-conventions)
- In light of the upcoming PRs that would affect *Favorites*, I would recommend to set this PR as Draft first to be updated once v1.9 goes out.

**Comment by deepali-hub** [comment]
I have updated tests as per above comments/suggestions.

File: __tests__/__theia__/extension.theia.test.ts
Comment (katelynienaber) [inline __tests__/__theia__/extension.theia.test.ts:20]: Why was this changed? Kinda nice to have all the constants use the same capitalization ^^' Maybe we could call it SHORTSLEEPTIME or something? idk

File: __tests__/__theia__/extension.theia.test.ts
Comment (katelynienaber) [inline __tests__/__theia__/extension.theia.test.ts:14]: Kinda prefer having the word Firefox in the variable name, as when I was reading the test I didn't know what FF was.

File: __tests__/__theia__/extension.theia.test.ts
Comment (katelynienaber) [inline __tests__/__theia__/extension.theia.test.ts:203]: Can we remove it?

File: src/theia/extension.theiaChrome.ts
Comment (katelynienaber) [inline src/theia/extension.theiaChrome.ts:18]: Can we have the same name for this 2 second sleep in both Chrome and Firefox test files?

File: src/theia/extension.theiaFF.ts
Comment (katelynienaber) [inline src/theia/extension.theiaFF.ts:19]: See other comments

**Review by katelynienaber — CHANGES_REQUESTED** [review]
This is awesome @deepali-hub! Just a few thoughts ^^'

File: __tests__/__theia__/extension.theia.test.ts
Comment (deepali-hub) [reply __tests__/__theia__/extension.theia.test.ts:203]: do you mean Line number 203 (commented code) should remove?

Comment (katelynienaber) [reply __tests__/__theia__/extension.theia.test.ts:203]: Yep! :)

File: __tests__/__theia__/extension.theia.test.ts
Comment (deepali-hub) [reply __tests__/__theia__/extension.theia.test.ts]: I have Changed variable name.

File: __tests__/__theia__/extension.theia.test.ts
Comment (deepali-hub) [reply __tests__/__theia__/extension.theia.test.ts]: I have mentioned Firefox now everywhere

File: __tests__/__theia__/extension.theia.test.ts
Comment (deepali-hub) [reply __tests__/__theia__/extension.theia.test.ts]: Removed commented block

File: src/theia/extension.theiaChrome.ts
Comment (deepali-hub) [reply src/theia/extension.theiaChrome.ts]: Gave same variable name in all the test files

File: src/theia/extension.theiaFF.ts
Comment (deepali-hub) [reply src/theia/extension.theiaFF.ts]: Gave same variable name in all the test files

**Review by katelynienaber — DISMISSED** [review]
LGTM :)

**Comment by deepali-hub** [comment]
Thank you for reviewing,  @katelynienaber

**Review by crawr — COMMENTED** [review]
You did a really good job with this, @deepali-hub ! Thanks for addressing my earlier comments.

Just some notes/questions
- I like the separate Locators, chrome and firefox files, but I think they should be moved from `src\theia` to `__tests__\__theia__` since we only use them for tests 
- After deleting the default profile from DATA SETS tree, do you think it is worth doing a negative test and check that it is no longer visible in the USS and JOBS trees?
- Overall, I think it should be quick to merge this in after the monorepo is merged since it is just 1 modified file and 3 new files.

**Comment by deepali-hub** [comment]
@crawr  I have worked on your notes->
1. Moved "theia" folder from src\theia to __tests__\__theia__
2. Covered negative test: I have added code to check that default profile is no longer visible in the USS and JOBS trees

File: packages/zowe-explorer/__tests__/__theia__/theia/extension.theiaFirefox.ts
Comment (crawr) [inline packages/zowe-explorer/__tests__/__theia__/theia/extension.theiaFirefox.ts:106]: These fields in the add profile dialog are dependent on the zosmf meta yaml file. If the tests are run locally on a different version of zowe cli, the additional/missing fields will cause the tests to fail. Perhaps there is a way the fields that are prompted could come from the meta file itself, rather than specific to the cli version.

**Review by crawr — APPROVED** [review]
Hey @deepali-hub , thanks for updating this. 
- Initially, I had a problem running the test locally since my CLI version is not the same as the one in the repo. As I mentioned in my other comment, there could be a better way to figure out what fields are prompted for the profile creation. But I think that could be handled in a separate PR.
- Another point of improvement that I saw was that perhaps there should be a cleanup routine before or after the tests. Running the tests as-is in the pipeline is fine since a new docker image is created per build. But locally, I ran the tests without the headless option and I saw that the creation of new profiles failed because the profile name is existing; but the tests passed because it was able to find the profile name in the tree. This could be an issue if the profile creation really failed.

I am fine with going with this PR for 1.12 and we can handle the improvements in a later version, especially since this is internal and transparent to the end user.

Just to document how I was able to run the tests locally:
1. Make sure my [geckodriver](https://github.com/mozilla/geckodriver/releases) and [chromedriver](https://sites.google.com/a/chromium.org/chromedriver/downloads) is updated and compatible with my browser versions
2. Update my zowe-cli version to `6.24.4` same as in the yarn.lock file
3. Delete the `zosmf_meta.yaml` and allow the new version of the cli to create a new one
3. Compile the project
4. Start a local Theia instance
5. Run the Theia tests

**Review by jellypuno — APPROVED** [review]
Looks good. tests are working fine

**Pull request merged**
